test/release/examples/primers/loops.chpl

```diff
@@ -0,0 +1,950 @@
+    b = -b;
+
+  writeln("After the forall loop over an array, B is: ", B);
+
+/*
+
+  Similarly, if we iterate over a domain or array that's distributed
+  across multiple locales with ``k`` cores each, each locale will tend
+  to use its ``k`` cores to iterate over the subset of the domain or
+  array that it owns locally:
+  
+*/
+
+  use BlockDist;
+
+  const BlockDom = Block.createDomain({1..n, 1..n});
+  var C: [BlockDom] real;
+
+  forall (i,j) in BlockDom do
+    C[i,j] = (100 * here.id) + i + j/1000.0;
+
+  writeln("After the forall loop over a distributed domain, C is:\n",C);
+
+  forall c in C do
+    c *= 2;
+
+  writeln("After the forall loop over a distributed array, C is:\n", C);
+
+/*
+
+  Because forall-loops invoke parallel iterators, the tasks they
+  create and where they run are not defined by the Chapel language,
+  but by the iterators themselves.  Any type supporting parallel
+  iteration should describe how its parallel iterators work as part of
+  its user-facing documentation.
+
+  Square Bracket Loops
+  --------------------
+
+  A third data-parallel loop form uses square brackets to define the
+  loop instead of the ``foreach`` or ``forall`` keywords.  For
+  example, such a loop may look like:
+
+*/
+  
+  [c in C] c -= 0.001;
+
+  writeln("After the first square bracket loop, C is:\n", C);
+
+/*
+
+  In this loop form, the square brackets can be thought of as
+  replacing the ``for[each|all]`` and ``do`` keywords, respectively.
+  This loop is both a shorthand for a data parallel loop, while also
+  supporting a "sliding scale" of parallelism.  Specifically, it will
+  be equivalent to a ``forall`` loop if its iterand has/is a parallel
+  iterator, and a ``foreach`` loop otherwise.
+
+  
+  Promotion and Data-Parallel Loops
+  ---------------------------------
+
+  In Chapel, an operator or procedure accepting a formal argument of
+  type ``t`` can be *promoted* by invoking the procedure with:
+
+  * an array whose elements are of type ``t``
+
+  * a range or domain whose indices are of type ``t``
+
+  Such promotions are equivalent to ``forall`` loops that iterate over
+  each of the promoted actual arguments in a zippered manner, passing
+  the respective elements into the operator or procedure.  For
+  example, given the procedure:
+
+*/
+
+  proc foo(ref x: real, t: (int, int), d: real) {
+    x = t(0) + t(1)/d;
+  }
+
+/*
+
+  The call:
+*/
+
+  foo(C, BlockDom, 100.0);
+
+  writeln("After the promoted call to foo(), C is:\n", C);
+
+/*
+
+  is equivalent to the forall-loop:
+
+*/
+
+  forall (c, ij) in zip(C, BlockDom) do
+    foo(c, ij, 100.0);
+
+  writeln("After the equivalent zippered forall loop, C is:\n", C);
+
+/*
+
+  As a result, the parallel calls to ``foo()`` will be executed using
+  the available processor cores on each of the locales that own a
+  portion of ``C`` since ``C`` is the first iterand in the `zip`
+  expression..
+  
+
+  A final note on data-parallel loops and legality / races
```

Comment:
This is important stuff that I would recommend putting after showing how all the loops iterate over arrays and before zippered iteration.